### ISSUE TITLE
Can this plugin be used on serial data (instead of page data)? - possibly by some workaround?

### ISSUE THREAD

**Issue by Chris75forumname** [cmacmackin/dokuwiki-plugin-structtask#5]
I like to keep it simle and I don't need a dedicated page for each task. All I need is email notifications on dates from the struct table.
Can this plugin be used with serial data entered from one single page or does it have to be page data? 
Or is there some workaround to use the plugin on serial data, like to manipulate the serial data to be perceived by the plugin as page data, even though it is entered trough a serial editor?

**Comment by cmacmackin** [comment]
I didn't design it to be able to work that way but there may be some way to adapt it. I will have to think about it. Do you have any experience working with PHP? I'd be happy to review a pull request. Otherwise I'm not sure when I'll get to this.

**Comment by Chris75forumname** [comment]
Yes, thank you for your quick response! If you think it is possible, I would look into this. Because I really like to have this feature. 

I really only need the CLI component for the email notification to look at the struct (serial) data and send out emails at 2a.m. to assignees according to due dates in the date fields. I don't need any of the edit and assignment related email action (that is triggered by a page save event).

If you could give me some guidance and where to look first. 

Am I right, in `cli.php` you extend the CLI Plugin with this code to create the notifiers directly from the struct data. So this CLI is run nightly by cronjob and creates and executes the due email actions, at the end of this function?
https://github.com/cmacmackin/dokuwiki-plugin-structtasks/blob/master/cli.php#L131

I think the problem is here: https://github.com/cmacmackin/dokuwiki-plugin-structtasks/blob/master/cli.php#L52, where it gets the tasks by `getPages()`. We would need to change this from pid to rid: https://www.dokuwiki.org/plugin:struct:dbinternals#data.

Maybe these are some things to look at, accessing serial data by rid:
https://codesearch.dokuwiki.org/xref/plugin/struct/meta/AccessTableSerial.php#14
https://github.com/cosmocode/dokuwiki-plugin-struct/blob/master/meta/Value.php#L68

Afterthought: Maybe we should do a spinoff of your plugin, called "structemailnotification", together.

**Comment by cmacmackin** [comment]
I definitely don't see a way to get this to work for the edit and assignment notifications. There is no event hook we could use for changes to global or serial data. Fortunately you've indicated that is not really a concern for you. (Although I feel like it would be useful to have such an event hook anyway. Perhaps worth proposing a PR on the struct plugin itself sometime...)

> Am I right, in cli.php you extend the CLI Plugin with this code to create the notifiers directly from the struct data. So this CLI is run nightly by cronjob and creates and executes the due email actions, at the end of this function?
>https://github.com/cmacmackin/dokuwiki-plugin-structtasks/blob/master/cli.php#L131
>
>I think the problem is here: https://github.com/cmacmackin/dokuwiki-plugin-structtasks/blob/master/cli.php#L52, where it gets the tasks by getPages(). We would need to change this from pid to rid: https://www.dokuwiki.org/plugin:struct:dbinternals#data.

That is indeed the bit that would need to change. I'm hesitant to start mucking about with `struct` internals, let alone directly accessing the database. The struct plugin says
> 3rd party developers should always interact with struct data through [this helper plugin](https://github.com/cosmocode/dokuwiki-plugin-struct/blob/master/helper.php#L17) only. If additional interface functionality is needed, it should be added here.

That helper plugin does not currently offer a way to get global or serial data, but I think it would be best to try adding the functionality there instead of here. I imaging a well-design and tested PR of this nature would be looked on favourably.

> Afterthought: Maybe we should do a spinoff of your plugin, called "structemailnotification", together.

I'm potentially open to this, but it would require some careful thought of in how to generalise it beyond this particular use case. That's not immediately obvious to me.

**Comment by Chris75forumname** [comment]
> There is no event hook we could use for changes to global or serial data.

Oh, I see! That's right! Is it because all of this is done via javascript?

> That helper plugin does not currently offer a way to get global or serial data, but I think it would be best to try adding the functionality there instead of here. I imaging a well-design and tested PR of this nature would be looked on favourably.

This is the access class for serial data (timestamp remains 0, rid is iterated through):
https://codesearch.dokuwiki.org/xref/plugin/struct/meta/AccessTableSerial.php#6

The man himself does it here in his plugin (he calls AccessTableSerial, extends it and uses it in his plugin):
https://codesearch.dokuwiki.org/xref/plugin/structpublish/meta/AccessTableStructpublish.php#29

and Szymon Olewniczak, unconventionally, makes use of the bureaucracy-struct interface ... :
https://codesearch.dokuwiki.org/xref/plugin/struct/helper/lookup.php#38
... for this in his plugin:
https://codesearch.dokuwiki.org/xref/plugin/structupdate/helper/update.php#53

... just something to check out.

> I'm potentially open to this, but it would require some careful thought of in how to generalise it beyond this particular use case. That's not immediately obvious to me.

Well, collaborative task management is one thing, email notification on duedates from a static table of repetitive duties is another thing. And the later has simpler requirements, it only relies on the CLI action. I need this for sceduling maintenance tasks. I will have to switch off all other notifications from structtasks plugin, except the cli notifications, because in a static task table, they make no sense. So, imo there are two separate usecases which would justify two different plugins.

We could have the user pass a list of schema names in the admin component for your plugin to look at. Then search each schema's complete set of data for all entries with a `rev == NULL && pid != NULL ` (no page data and no global data) and iterate through all rids. https://www.dokuwiki.org/plugin:struct:dbinternals#data

**Comment by Chris75forumname** [comment]
Finally, I have found my workaround. For now, I put together a quick setup for "dummy" page creation by using the addnewpage plugin and some namespace template with `strftime` date parameters. I create dummy pages in a namespace "struct" named with a timestamp, that I insert using javascript in `/conf/userscript.js`. This way I create the pages for storage of the page data for structtasks on-the-fly and redirect back to the aggregation page, then, automatically. So I preserve the feel of a serial aggregation editor, as much as possible. 

So I consider this OP as solved, for now. Cheers! 👍
